### DISCUSSION TITLE
Proposal: Tuple deconstruction in lambda argument list

### DISCUSSION THREAD

**Discussion by quinmars** [dotnet/csharplang#258] [General]
A lambda that accepts a (single) tuple as argument can only be defined with the tuple as a single argument and not by the tuples items. I’d like to see that the tuple deconstruction also works in the argument list of lambdas, as shown in the following examples:

```csharp
Func<(int, int), int> f1 = (t) => t.Item1 + t.Item2; // works today
Func<(int, int), int> f2 = ((a, b)) => a + b;       // should work
Func<(int, int), int> f3 = t => t.Item1 + t.Item2;  // works today
Func<(int, int), int> f4 = (a, b) => a + b;         // would be nice

```

Note that in the last line the parenthesis belong to the tuple deconstruction and not the lambda argument list.

This feature would be handy for Linq or Rx method chaining:
```csharp
var r = Enumerable
        .Range(1, 10)
        .Select(i => (i + 1, i * i))
        .Where((a, b) => 2 * a < b)
        .OrderBy((a, b) => b)
        .Last();
```

Of course the given calculation doesn’t make much sense, but I guess you get the idea. That’s why I'd like to see this feature. But to be consequent, more complex constructs would become possible:

```csharp
Func<(int, int), (int, int), int> f2 = ((a, b), (c, d)) => a + b + c + d;
```

And probably that should also be extended to methods (and delegates):

```csharp
void Do((int x, int y), string name)
{
    // use here x, y, and name
}
```

**Comment by DavidArno** [comment]
I agree that:
```cs
Func<(int, int), int> f2 = ((a, b)) => a + b;
```
should work already. This is a C# 7 bug in my view.

However, I also agree that it would be good to have implicit deconstruction of tuples to parameter lists as an added bonus, as the need to use `((x, y))` seems clunky.

**Comment by ufcpp** [comment]
IMO, this should work in the near future, but not now. This may be part of [Pattern Matching](https://github.com/dotnet/csharplang/issues/45) in the [7.X milestone](https://github.com/dotnet/csharplang/milestone/7).

```cs
var t = (1, 2);

if (t is (int a, int b)) // does not work today, but will work in 7.X
{
}

switch (t)
{
    case (int a, int b): // does not work today, but will work in 7.X
        break;
}

Func<(int, int), int> f = ((a, b)) => a + b; // does not work today
```

**Comment by Thaina** [comment]
I would agree only on

`Func<(int, int), int> f2 = ((a, b)) => a + b;`

But I don't agree with implicit deconstruction. In the future might be able to do anonymous function so we should always explicit. Should always avoid ambiguous syntax

**Comment by asdfgasdfsafgsdfa** [comment]
I just stumbled about this limitation as well (the second case, but I agree with all)
I wonder why it is that way. Maybe it's that people just didn't think of it, which would be no big deal.
But I wonder if there are any real reasons.

**Comment by HaloFour** [comment]
Some of the explicitly mentioned LINQ-based examples would be ambiguous.  Those LINQ methods have overloads which take delegates that accept two parameters.  If you could deconstruct a tuple in the lambda parameters without a special syntax it wouldn't be possible for the compiler to disambiguate between [`Enumerable.Where(IEnumerable<(T1, T2)>, Func<(T1, T2), bool>)`](https://msdn.microsoft.com/en-us/library/bb534803(v=vs.110).aspx) or [`Enumerable.Where(IEnumerable<(T1, T2)>, Func<(T1, T2), int, bool>)`](https://msdn.microsoft.com/en-us/library/bb549418(v=vs.110).aspx)

**Comment by solvingj** [comment]
Great suggestion, but Kudos to @HaloFour for identifying the problem with the proposed syntax. 

I see that this is kept open as a proposal, but was a new issue opened for the bug with the case below?

`Func<(int, int), int> f2 = ((a, b)) => a + b;`

**Comment by jcouv** [comment]
For reference, linking to LDM notes: https://github.com/dotnet/csharplang/blob/master/meetings/2017/LDM-2017-03-15.md#deconstruction-in-lambda-parameters

**Comment by jcouv** [comment]
I want to point out the relationship of this proposal to another proposal (splatting).
There are two ways of looking at the same problem: passing a composite argument to an invocation and make its parts available as parts in the method code.

 - One is [splatting of tuples into arguments](https://github.com/dotnet/csharplang/issues/424): calling a method with a tuple, and have the elements of the tuple splat into distinct arguments of the method. You can do this today manually by invoking `M(tuple.first, tuple.second)`.
 - The other (described by present proposal) is deconstruction in lambda parameters: when a lambda is invoked with a parameter, deconstruct that parameters into elements and use those elements in the lambda body. You can do this today manually by defining a lambda as `x => { var (first, second) = x; Write(first); Write(second); }`.

**Comment by Thaina** [comment]
@jcouv To have lambda be a one line return without `return` keyword is important at least for me

```C#
var point3s = new[] { (0,0,0),(1,1,1) }
var point2s = point3s.Select(((x,y,z)) => (x,y));
```

**Comment by Timwi** [comment]
This should be implemented for lambdas and methods. Thanks for writing up the proposal.

I would like to point out that the OP has incorrect/inconsistent syntax for the case of methods. OP wrote:

```cs
void Do((int x, int y), string name)
{
    // use here x, y, and name
}
```

However, this shouldn't compile because `(int x, int y)` is a valid type name, so it's missing the parameter name. The following two syntaxes should both work in order to be consistent with variable declaration syntax and pattern matching:

```cs
void Do((int x, int y) (x, y), string name) { }
void Do((int, int) (x, y), string name) { }
```

**Comment by jnm2** [comment]
@Timwi 

> it's missing the parameter name

Just curious: what would the actual parameter name end up being, for older compilers?

**Comment by Timwi** [comment]
Good point, I hadn't considered that. It's actually not just older compilers that are affected: you need to be able to use named arguments to pass a tuple-typed argument to a method with optional parameters.

My proposal here would be that the syntax could be extended to accommodate an extra parameter name, for example:

```cs
public void Method((int, int) (x, y) point) { }
```

but this would be optional; if such an extra parameter name is not provided, the compiler would generate one by combining the names from the deconstruction syntax. For example, `(int, int) (index, count)` could lead to the default parameter name of `indexCount`. It would then have to throw a new type of error message if another provided parameter name clashes with that.

**Comment by Thaina** [comment]
Why this feature not be championed for C# 8.0 I wonder

**Comment by jhudsoncedaron** [comment]
I have to admit I expected it to work. I typed

    .Select((a, b) => { /* 4 lines of code here */ return something; })

and found out it didn't work.

**Comment by Thaina** [comment]
@jhudsoncedaron Because that was the actual syntax for normal lambda

Even this feature work it should be

```C#
.Select(((a,b)) => { /* 4 lines of code here */ return something; })
```

I think we should deprecate the no parentheses syntax for lambda

**Comment by CyrusNajmabadi** [comment]
> I think we should deprecate the no parentheses syntax for lambda

Wouldn't that break virtually 100% of all projects out there?  Or did you mean a specific non-common form?

**Comment by Thaina** [comment]
@CyrusNajmabadi I mean just mark it as deprecation to pop the warning just for discourage it usage of the `.Select(x => x)`

We might not able to remove it for all eternity but it really confusing things up

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi I mean just mark it as deprecation to pop the warning just for discourage it usage of the .Select(x => x)

You would be deprecating an incredibly common, well understood, non-problematic pattern for the ecosystem across pretty much every codebase...

The only way this could ever be ok would be to have the biggest benefit ever coming to the language.  It would literally have to be as good, or better, than linq itself ot warrant that sort of deprecation.

**Comment by Thaina** [comment]
@CyrusNajmabadi I think it really better have non ambiguous pattern between `.Select(x => x)` and `.Select((x) => x)` that just mean the same thing. It just save 2 parentheses character but then it cause confusion to tuple. Now it make less obvious is it direct parameter or tuple deconstruction

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi I think it really better have non ambiguous pattern between .Select(x => x) and .Select((x) => x)

I can't possibly see how that scenario would be worth breaking nearly every codebase in existence.

> It just save 2 parentheses character but then it cause confusion to tuple.

Tuples are orders of magnitude less important than Linq and Lambdas :)  They're also way less important than *not breaking all the c# projects in existence* :) 

> Now it make less obvious is it direct parameter or tuple deconstruction

This is an incredibly marginal scenario. And not at all important enough to deprecate the very widely used, understood, and liked syntax for a simple 1-arg lambda.

**Comment by Thaina** [comment]
@CyrusNajmabadi As I said just mark deprecation for discourage by warning does not mean it would be removed anytime soon. It does not break any existing codebase

**Comment by Thaina** [comment]
@CyrusNajmabadi Lambdas  is important but `.Select(x => x)` is not any important because we already have `.Select((x) => x)` and we really should not get rid of those parentheses from the start

**Comment by theunrepentantgeek** [comment]
@Thaina it does break existing code because there are a **lot** of existing codebases that run with _warnings-as-errors_ turned on.

It would also add noise to perhaps the most common use of lambda expressions, all without providing more than a jot of value.

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi As I said just mark deprecation for discourage by warning does not mean it would be removed anytime soon. It does not break any existing codebase

It breaks all codebases that have 'warn as errors' on.  It also means producing *thousands* of warnings for codebases.  I just checked Roslyn itself.  This would generate over *15,000* warnings.  It would warn on literally one out of every three files in the codebase! :-/

> @CyrusNajmabadi Lambdas is important but .Select(x => x) is not any important

It's important because it's how people are used to typing lambdas and what is idiomatic across the c# ecosystem.  Its like saying "we can just deprecate ```+``` on strings, since you can just call .Concat yourself".  It's not happening.  You can't take commonly used constructs that have existed for more than a decade, and just deprecate them.

>  we already have .Select((x) => x) and we really should not get rid of those parentheses from the start

That ship has *long* sailed.  The decisions on how Lambdas could be written were made by Anders and completely accepted by teh C# team and the entire ecosystem.  It's not an area where there is a problem.  It's not an area where there is very minimal usage (like, say, anonymous-delegates `delegate () {}`).  You're talking about one of the most common patterns out there.  Again, it's akin to saying "we're going to deprecate referring to an instance field with `foo` since you can write `this.foo`".  It's *not* going to happen.

**Comment by Thaina** [comment]
@CyrusNajmabadi 

> That ship has long sailed

Every ship sailed could be _discourage_ from using when we explicitly said that we should use another ship. Like we used to have `async callback` and `IAsyncResult` but now we have `Task` and `async/await` instead so the former was discourage from being used even though it still around for backward compatility

Sure I would not agree on breaking anything. If just warning still break then I would retract my request but the purpose is we should do anything to discourage this pattern and encourage people to use full pattern for lessen ambiguous with lambda tuple deconstruction

**Comment by HaloFour** [comment]
@Thaina 

Also note that you are in an extreme minority of people who would want to see those parenthesis be mandatory.  The ability to omit those parenthesis is desired and idiomatic.  Even if this wasn't a massive breaking change (which it is), nobody would want this.

**Comment by Thaina** [comment]
@HaloFour I am quite fine with it all along but now it seem hindrance to the tuple deconstruction in lambda syntax

**Comment by jhudsoncedaron** [comment]
I don't care about the lambda parenthesis debate. I always place the parenthesis around lambda arguments but it doesn't matter much in this case. I tried it both ways before concluding that C# can't do tuple deconstruction there.

**Comment by kofifus** [comment]
Seems like this is not on the map for C# 8.0 as well ? any updated on this proposal ?

**Comment by gafter** [comment]
@kofifus I'm not aware of any LDM champion for this idea.

**Comment by ygoe** [comment]
I'm wondering why the double-parentheses syntax doesn't work. This bug makes using language tuples in lambda functions unusable. I can only use Item1 and Item2 which says nothing. When I try to write this code (which feels correct, based on my understanding of tuple syntax elsewhere) today, the compiler freaks out completely, giving me several errors, some duplicates, in multiple places, leaving me with no clue at all what the problem is. This should be given a higher priority. I'd rather see bugs like this fixed before anybody gets crazy stuff like non-nullable nullable reference types.

**Comment by CyrusNajmabadi** [comment]
> I'm wondering why the double-parentheses syntax doesn't work.

Because it's not part of the language.

>  This bug makes using language tuples in lambda functions unusable

Clearly not.  Lots of people use them all the time.

> I can only use Item1 and Item2 which says nothing. 

Not the case.  You can write:

`Foo(((int first, int second) t) => ... t.first ... t.second ...)`

Now, this isn't wonderful.  And languages like javascript/typescript make this nicer like so: `Foo(([first, second]) => ... first ... second ...)`, allowing you to avoid naming the tuple instance. 

So, it's def the case this could be nicer.  But it's not the case that things are unusable.

**Comment by nblumhardt** [comment]
> Not the case. You can write:
> Foo(((int first, int second) t) => ... t.first ... t.second ...)

Thanks for pointing that out, @CyrusNajmabadi 👍 

(Unfortunately for me it was more like: "... _search the web and read Stack Overflow and GitHub tickets for half an hour_, and then write: ..." 😄)

**Comment by kswoll** [comment]
The solution @CyrusNajmabadi proposed works, but if you move the naming over to where the tuple is constructed, it's quite a bit nicer (particularly if you're returning `IEnumerable<TUPLE>` out of a method and name the tuple values in the return type):

```
var r = Enumerable
    .Range(1, 10)
    .Select(i => (a: i + 1, b: i * i))
    .Where(x => 2 * x.a < x.b)
    .OrderBy(x => x.b)
    .Last()
```

**Comment by nblumhardt** [comment]
Thanks, @kswoll . 

In my case the tuple returned generically from an API that can't give the fields meaningful names (only the consumer of the API knows what each field in the tuple represents), but the tip is appreciated!

**Comment by jhudsoncedaron** [comment]
My case was similar to nbumhardt's case. The tuple appeared in template parameter expansion.

**Comment by Thaina** [comment]
Sadly this feature was so stale

**Comment by Lonli-Lokli** [comment]
Will it be implemented in after-covid era?

**Comment by jhudsoncedaron** [comment]
After we got named ValueTuple elements I stopped caring.

**Comment by Thaina** [comment]
If anyone would champion this in the future can it moved back to issue board?

**Comment by EduardSergeev** [comment]
Another workaround is to use [switch expression](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/switch-expression) added to C# 8:

```c#
var r = Enumerable
    .Range(1, 10)
    .Select(i => (i + 1, i * i))
    .Where(_ => _ switch { var (a, b) => 2 * a < b })
    .OrderBy(_ => _ switch { var (a, b) => b })
    .Last();
```
and
```c#
Func<(int, int), int> f2 = _ => _ switch { var (a, b) => a + b };
```

**Reply by DavidArno** [reply]
Surely that is more clearly written as:
```c#
var r = Enumerable
    .Range(1, 10)
    .Select(i => (a:i + 1, b:i * i))
    .Where(ab => 2 * ab.a < ab.b)
    .OrderBy(ab => ab.b)
    .Last();
```

**Comment by alrz** [comment]
> `Func<(int, int), int> f2 = ((a, b)) => a + b;       // should work`

How about `Func<(int, int), int> f2 = var (a, b) => a + b;` that's what deconstruction looks like everywhere.

I would add `M(out var (a, b))` to this same wishlist.

I think none of these should contribute to overload resolution or type inference though - deconstruction conversion would apply as the last step and may fail.

**Reply by alrz** [reply]
Looks like LDT has already foresaw the possibility of this usage.

> The contextual keyword 'var' cannot be used as an explicit lambda return type

**Reply by 333fred** [reply]
I'll be honest: I don't think we foresaw anything there. I think we just didn't want to allow the return type to be explicitly specified as var.

**Comment by garcipat** [comment]
I was looking for that. Wrote this extension to make that possible. But I would appreciate if this would be standard: 
```csharp
public static IEnumerable<TResult> Select<TItem1, TItem2, TResult>(this IEnumerable<(TItem1, TItem2)> source, Func<TItem1, TItem2, TResult> selector)
        {
            return source.Select(s => selector(s.Item1, s.Item2));
        }
```

**Comment by CyrusNajmabadi** [comment]
I will champion this.

**Reply by Thaina** [reply]
Are there any milestone for this feature?

**Reply by CyrusNajmabadi** [reply]
> Are there any milestone for this feature?

There is not

**Comment by wheatandtares** [comment]
@CyrusNajmabadi Just want to surface this to your attention again. I use higher order functions and tuples very frequently, so this syntax (only basic tuple deconstruction in lambda parameter list) is by far my most desired feature.

```
Func<(int, int), int> f2 = ((a, b)) => a + b;
```

I find myself being shaken out of my flow and writing boilerplate like this constantly.
```
Func<(int, int), int> f2 = ab =>
{
  var (a, b) = ab;
  return a + b;
};
```

Given C#'s mission of reducing boilerplate, how simple and intuitive of a language change this would be, and how cohesive it is with recent changes to the language, I would ask that we continue to advocate for this proposal :pray: 

Thanks for all ya'll do! C# continues to be the best language on the market for those that rate it on its objective merits.

P.S. for what it's worth, I think we should NOT add support for 
```
Func<(int, int), int> f2 = (a, b) => a + b;
```
as it is ambiguous syntax, doesn't make the code more clear, and creates "magic" to confuse new programmers.

**Comment by jnm2** [comment]
There's now a proposal at <https://github.com/dotnet/csharplang/issues/9848> and @CyrusNajmabadi and I will bring this to an upcoming LDM as the schedule allows!

**Comment by Thaina** [comment]
7 years of pain
